### PR TITLE
feat(helmchart): add extraContainers to speaker

### PR DESCRIPTION
**Pull request by DrBu7cher**
In the environment we are using we would like to deploy a container right next to every MetalLB Speaker. The easiest and most reliable way for us would be to use the functionality this PR would add.

If you want me to add this to the `controller.yaml` too, feel free to tell me.

### CODE REVIEW

**Comment by fedepaol** [comment]
> In the environment we are using we would like to deploy a container right next to every MetalLB Speaker. The easiest and most reliable way for us would be to use the functionality this PR would add.
> 
> If you want me to add this to the `controller.yaml` too, feel free to tell me.

Can you expand a bit the use case, to understand why and how it could help other metallb users.

**Comment by DrBu7cher** [comment]
Of course.
We have a container that is going to configure the announced IPs on the interface of the nodes that metallb is announcing from. We could create a seperate DaemonSet with the same configuration parallel to metallb, but it would be easier and more reliable to have the container be part of the speaker-pod(s).

**Comment by fedepaol** [comment]
> Of course. We have a container that is going to configure the announced IPs on the interface of the nodes that metallb is announcing from. We could create a seperate DaemonSet with the same configuration parallel to metallb, but it would be easier and more reliable to have the container be part of the speaker-pod(s).

And what's the reason again (sorry if I insist, trying to understand if it makes sense as a metallb feature)? 
This is for l2 I guess, as you can't have multiple nodes with the same ip. Also, this goes under the assumption that each node has only one interface?
And also also, what is the relationship between the extra container and the speaker? It is just for deployment commodity I think, as it won't be able to inspect what ips are being announced from a given node.

**Comment by DrBu7cher** [comment]
If you ping the BGP IP, the ICMP packets go into a loop between node and router and you never get a response.
This is working as expected from your side, as far as I know,
> Also, please be aware that pinging the service IP won’t work.
[source](https://metallb.universe.tf/troubleshooting/#general-concepts)
but we want our IPs to be pingable.

We configured/added the IPs to the node interfaces on the host that metallb is running on and ping works now. I built a container/controller that configures the IP on the host, so we can deploy and manage it through Kubernetes.

It can handle multiple interfaces, but since I only care about ping, it would not be a problem if I would configure the IP on the loopback interface always.

**Comment by fedepaol** [comment]
Please let's add this to the controller too, for symmetry. I am not sure I like it but it is nearly to not invasive.
Also, CI will complain about the helm docs I think

cc @oribon

File: charts/metallb/values.yaml [hunk 1]
```diff
@@ -249,6 +249,8 @@ controller:
     successThreshold: 1
     timeoutSeconds: 1
 
+  extraContainers: {}
```
Comment (oribon) [inline charts/metallb/values.yaml:252]: shouldn't this be an array?

**Review by oribon — COMMENTED** [review]
specific use-case aside, I think this parameter is common enough for us to include as well (and shouldn't cause any harm I guess) and +1 for having it in the controller also.

File: charts/metallb/values.yaml (hunk 1 above)
Comment (DrBu7cher) [reply charts/metallb/values.yaml:252]: You are correct.

Comment (DrBu7cher) [reply charts/metallb/values.yaml]: Fixed

**Comment by DrBu7cher** [comment]
Please approve it again :rocket:

**Comment by oribon** [comment]
sorry I missed it, mind adding an array validation in `values.schema.json` before?

**Comment by DrBu7cher** [comment]
Can you check again? Is this what you meant?
Is something else missing?

**Comment by oribon** [comment]
> How can I test if it is correct?

by changing the value to something that is not an array:
`extraContainers: "abc"`
you get:
```
$ helm template .
Error: values don't meet the specifications of the schema(s) in the following chart(s):
metallb:
- speaker.extraContainers: Invalid type. Expected: array, given: string
- speaker: Must validate all the schemas (allOf
```

**Comment by DrBu7cher** [comment]
The `values.scheme.json` was not correct.
Please approve again @oribon

**Comment by DrBu7cher** [comment]
Like so?

**Comment by oribon** [comment]
close :) put the `controller placement` commit first and all the other changes in another commit (so there are 2 commits in total)

**Comment by DrBu7cher** [comment]
Please approve @oribon

**Pull request merged**